### PR TITLE
Enforce rustdoc-gui test-suite run

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Part of https://github.com/rust-lang/rust/issues/84550

### CODE REVIEW

**Comment by jyn514** [comment]
> By the way, Is there something else to do alongside this to be sure that the rustdoc-gui test-suite will be run on linux targets?

Yes, make sure nodejs and npm are installed on those builders. You can check by looking at the build logs for `x86_64-gnu-llvm-10` and making sure `rustdoc-gui` isn't skipped.

File: src/bootstrap/test.rs [hunk 1]
```diff
@@ -773,7 +773,9 @@ impl Step for RustdocGUI {
     const ONLY_HOSTS: bool = true;
 
     fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
-        run.path("src/test/rustdoc-gui")
+        let builder = run.builder;
+        let run = run.path("src/test/rustdoc-gui");
+        run.default_condition(builder.config.nodejs.is_some() && builder.config.npm.is_some())
```
Comment (jyn514) [inline src/bootstrap/test.rs:778]: You should also make the "warning: npm is missing" line a hard error, or this doesn't help with making sure the test is run.

Comment (GuillaumeGomez) [reply src/bootstrap/test.rs:778]: True. Updating.

**Comment by GuillaumeGomez** [comment]
Well, I guess now I have to add it in multiple places then. :3

**Comment by Mark-Simulacrum** [comment]
Why are we putting this check on the PR builder? I understand conceptually that it's nice to get breakage early, but I worry that we shouldn't just add onto that builder without consideration. It may be worthwhile to take the approach of the tools builder, where we only run it directly on PRs with modification to the tools directory (while bors full builds always run it), and move the check from the aux builder to the tools builder.

**Comment by jyn514** [comment]
> It may be worthwhile to take the approach of the tools builder, where we only run it directly on PRs with modification to the tools directory (while bors full builds always run it), and move the check from the aux builder to the tools builder.

Changes to rustdoc almost never modify the tools directory, so that means this check won't run on PRs.

**Comment by Mark-Simulacrum** [comment]
Sure, we can add rustdoc to the list of things which trigger a build?

**Comment by GuillaumeGomez** [comment]
> Sure, we can add rustdoc to the list of things which trigger a build?

Just to be sure: it's something to do on "infra side", not in this PR, right?

**Comment by jyn514** [comment]
@GuillaumeGomez no, it should go in this PR. You should find where `x86_64-gnu-tools` determines if the check should be skipped and modify it to take src/librustdoc into account.

**Comment by GuillaumeGomez** [comment]
I see, thanks for the clarification.

File: src/ci/docker/host-x86_64/x86_64-gnu-llvm-10/Dockerfile [hunk 2]
```diff
@@ -19,8 +19,10 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
   libssl-dev \
   pkg-config \
   zlib1g-dev \
-  xz-utils \
-  nodejs
+  xz-utils
+
+RUN curl -sL https://nodejs.org/dist/v14.4.0/node-v14.4.0-linux-x64.tar.xz | tar -xJ
+ENV PATH="/node-v14.4.0-linux-x64/bin:${PATH}"
```
Comment (jyn514) [inline src/ci/docker/host-x86_64/x86_64-gnu-llvm-10/Dockerfile:25]: Is npm not available through apt?

Comment (GuillaumeGomez) [reply src/ci/docker/host-x86_64/x86_64-gnu-llvm-10/Dockerfile:25]: Apparently not based on the previous error.

**Comment by GuillaumeGomez** [comment]
Another thing to be noted: if the `browser-ui-test` package isn't installed, we skip the rustdoc-gui tests. Should I also add the installation of this package in the `Dockerfile` I updated so the tests are run?

**Comment by jyn514** [comment]
> if the browser-ui-test package isn't installed, we skip the rustdoc-gui tests

I thought you changed it to give a hard error? If not, can you do that now?

**Comment by GuillaumeGomez** [comment]
No, I only enforced the presence of `npm`. Adding it.

**Comment by GuillaumeGomez** [comment]
It seems to be working now. I was expecting a failure coming from the run of the rustdoc-gui tests though... Did I miss something?

**Comment by jyn514** [comment]
@GuillaumeGomez you didn't fix the original issue - you need to move the `RUST_CHECK_TARGET` from `x86_64-gnu-aux` to `x86_64-gnu-llvm-10`: https://github.com/rust-lang/rust/issues/84550#issuecomment-826333297

**Comment by jyn514** [comment]
Err I mean `x86_64-gnu-tools` now that it checks for changes to librustdoc.

File: src/ci/docker/host-x86_64/x86_64-gnu-tools/checktools.sh [hunk 3]
```diff
@@ -17,6 +17,7 @@ python3 "$X_PY" test --stage 2 --no-fail-fast \
     src/tools/rls \
     src/tools/rustfmt \
     src/tools/miri \
+    src/librustdoc
```
Comment (jyn514) [inline src/ci/docker/host-x86_64/x86_64-gnu-tools/checktools.sh:20]: This should also check for changes to the testsuite (like the ones in this PR).
```suggestion
    src/librustdoc src/rustdoc-json-types src/test/rustdoc{,-ui,-js,-gui,-json,-js-std}
```
(and you'll also need to change the shebang to #!/bin/bash to support globbing)

Comment (jyn514) [reply src/ci/docker/host-x86_64/x86_64-gnu-tools/checktools.sh:20]: Oh wait hold on, this isn't the right thing to modify - this is an `x.py test` command. You want to update the logic that causes this to run in the first place - there's a comment about TOOLSTATE_FILE, maybe search for that.

Comment (GuillaumeGomez) [reply src/ci/docker/host-x86_64/x86_64-gnu-tools/checktools.sh:20]: I think it refers to `src/bootstrap/toolstate.rs` which has `STABLE_TOOLS` and `NIGHTLY_TOOLS` arrays that contains lists somewhat similar. However, I don't see how it relates to this... And I'm not even sure how this helps in our case...

**Comment by GuillaumeGomez** [comment]
Strangely enough, installing the npm package doesn't install the package. So I'm pretty sure I'm missing something obvious here...

File: src/ci/docker/host-x86_64/dist-various-2/Dockerfile [hunk 4]
```diff
@@ -29,6 +28,12 @@ RUN apt-get update && apt-get build-dep -y clang llvm && apt-get install -y --no
 RUN apt-key adv --batch --yes --keyserver keyserver.ubuntu.com --recv-keys 74DA7924C5513486
 RUN add-apt-repository -y 'deb http://apt.dilos.org/dilos dilos2 main'
 
+# Install required dependencies from browser-UI-test framework
+# For now, we need to use `--unsafe-perm=true` to go around an issue when npm tries
+# to create a new folder. For reference:
+# https://github.com/puppeteer/puppeteer/issues/375
+RUN npm install browser-ui-test -g --unsafe-perm=true
```
Comment (Mark-Simulacrum) [inline src/ci/docker/host-x86_64/dist-various-2/Dockerfile:35]: Why is anything changing in dist-various-2? This builder shouldn't even run tests?

File: src/ci/docker/host-x86_64/x86_64-gnu-aux/Dockerfile [hunk 5]
```diff
@@ -30,20 +30,10 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
...
-RUN npm install browser-ui-test -g --unsafe-perm=true
-
 COPY scripts/sccache.sh /scripts/
 RUN sh /scripts/sccache.sh
 
 COPY scripts/cmake.sh /scripts/
 RUN /scripts/cmake.sh
 
 ENV RUST_CONFIGURE_ARGS --build=x86_64-unknown-linux-gnu
-ENV RUST_CHECK_TARGET check-aux-and-gui
```
Comment (Mark-Simulacrum) [inline src/ci/docker/host-x86_64/x86_64-gnu-aux/Dockerfile:49]: This should move back to check-aux, not be removed entirely.

File: src/ci/docker/host-x86_64/x86_64-gnu-llvm-10/Dockerfile [hunk 6]
```diff
@@ -32,6 +31,17 @@ ENV RUST_CONFIGURE_ARGS \
...
+
+# Install required dependencies from browser-UI-test framework
+# For now, we need to use `--unsafe-perm=true` to go around an issue when npm tries
+# to create a new folder. For reference:
+# https://github.com/puppeteer/puppeteer/issues/375
+RUN npm install browser-ui-test -g --unsafe-perm=true
+
+ENV RUST_CHECK_TARGET check-aux-and-gui
```
Comment (Mark-Simulacrum) [inline src/ci/docker/host-x86_64/x86_64-gnu-llvm-10/Dockerfile:43]: We discussed placing the new check in the tools builder, not here?

File: src/ci/docker/host-x86_64/x86_64-gnu-tools/Dockerfile [hunk 7]
```diff
@@ -21,6 +21,15 @@ RUN sh /scripts/sccache.sh
...
+
+# Install required dependencies from browser-UI-test framework
+# For now, we need to use `--unsafe-perm=true` to go around an issue when npm tries
+# to create a new folder. For reference:
+# https://github.com/puppeteer/puppeteer/issues/375
+RUN npm install browser-ui-test -g --unsafe-perm=true
+
 COPY host-x86_64/x86_64-gnu-tools/checktools.sh /tmp/
 
 ENV RUST_CONFIGURE_ARGS \
```
Comment (Mark-Simulacrum) [inline src/ci/docker/host-x86_64/x86_64-gnu-tools/Dockerfile:35]: We'll want to:

* remove the check-aux-and-gui Makefile rule
* add to the SCRIPT here an explicit invocation of the src/test/rustdoc-gui test, likely after we call checktools.sh (via &&)

**Comment by Mark-Simulacrum** [comment]
With regards to the reason this isn't working, I don't really know enough about npm to debug why the successful installation isn't getting picked up later on when you check.

This PR is also missing the "run tools builder on changes to rustdoc" rule here - https://github.com/rust-lang/rust/blob/c488f15700f15664b3bbda742c3f53f2b7a508d9/src/ci/scripts/should-skip-this.sh#L17

We can likely include a check for if *that* file changes, then run the builder, as well to enable testing in this PR.

File: src/bootstrap/test.rs [hunk 8]
```diff
@@ -802,53 +804,52 @@ impl Step for RustdocGUI {
...
-            command.arg("src/test/rustdoc-gui/lib.rs").arg("-o").arg(&out_dir);
-            builder.run(&mut command);
+        builder.ensure(compile::Std { compiler: self.compiler, target: self.target });
 
-            for file in fs::read_dir("src/test/rustdoc-gui").unwrap() {
-                let file = file.unwrap();
-                let file_path = file.path();
-                let file_name = file.file_name();
+        // The goal here is to check if the necessary packages are installed, and if not, we
+        // display a warning and move on.
```
Comment (jyn514) [inline src/bootstrap/test.rs:813]: This is no longer accurate, it gives a hard error now.

File: src/bootstrap/test.rs [hunk 9]
```diff
@@ -802,53 +804,52 @@ impl Step for RustdocGUI {
...
+        let lines = command
+            .output()
+            .map(|output| String::from_utf8_lossy(&output.stdout).to_string())
+            .unwrap_or(String::new());
+        if !lines.contains(&" browser-ui-test@") {
+            println!(
+                "warning: rustdoc-gui test suite cannot be run because npm `browser-ui-test` \
+                 dependency is missing",
+            );
+            println!(
```
Comment (jyn514) [inline src/bootstrap/test.rs:825]: This should print to stderr I think. Also "warning" is no longer accurate.
```suggestion
            eprintln!(
                "error: rustdoc-gui test suite cannot be run because npm `browser-ui-test` \
                 dependency is missing",
            );
            eprintln!(
```

File: src/ci/docker/host-x86_64/x86_64-gnu-llvm-10/Dockerfile [hunk 10]
```diff
@@ -19,13 +19,11 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
...
   pkg-config \
   zlib1g-dev \
-  xz-utils \
-  nodejs
+  xz-utils
 
 COPY scripts/sccache.sh /scripts/
 RUN sh /scripts/sccache.sh
 
-# using llvm-link-shared due to libffi issues -- see #34486
```
Comment (jyn514) [inline src/ci/docker/host-x86_64/x86_64-gnu-llvm-10/Dockerfile:28]: This comment should not have been removed.

File: src/test/rustdoc-gui/basic.goml [hunk 11]
```diff
@@ -1,4 +1,4 @@
 goto: file://|DOC_PATH|/index.html
-assert: ("#functions")
+assert: ("#functi")
```
Comment (jyn514) [inline src/test/rustdoc-gui/basic.goml:2]: Why did this change?

Comment (GuillaumeGomez) [reply src/test/rustdoc-gui/basic.goml:2]: It's part of my second commit "TO BE REMOVED". I want to be sure the CI is failing because this test failed. :)

**Comment by GuillaumeGomez** [comment]
I might need a bit of help for npm dependency not being installed: if I understand correctly, `ENV SCRIPT` is setting the env variable `SCRIPT` value by giving the scripts to execute, but they're executed outside of this Dockerfile. But if this is the case, where are they run?

**Comment by jyn514** [comment]
@GuillaumeGomez make sure you switch the labels when you ask for help, or it might get lost in the sea of pings.

**Comment by GuillaumeGomez** [comment]
@jyn514 Indeed, thanks!

File: src/ci/docker/host-x86_64/x86_64-gnu-llvm-10/Dockerfile [hunk 12]
```diff
@@ -19,8 +19,7 @@ RUN apt-get update && apt-get install -y --no-install-recommends \
   libssl-dev \
   pkg-config \
   zlib1g-dev \
-  xz-utils \
-  nodejs
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/ci/docker/host-x86_64/x86_64-gnu-llvm-10/Dockerfile:23]: Because of the changes, if nodejs is present, the tests will fail if the rustdoc-gui test suite cannot be run.

[the remark this answers is not in the document]
Comment (Mark-Simulacrum) [reply src/ci/docker/host-x86_64/x86_64-gnu-llvm-10/Dockerfile:23]: Hmm. Not wanting to install browser-ui-test in particular seems OK though -- I think gating on just nodejs may have been too little. Can we gate on having browser-ui-test installed, and skip the build *by default* if it isn't? That way it should still be OK to have nodejs installed in builders, but we'll just need to explicitly ask for the browser UI tests in one of the builders that has it preinstalled.

[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/ci/docker/host-x86_64/x86_64-gnu-llvm-10/Dockerfile:23]: Well, we're worried that the rustdoc-gui test suite gets silently ignored once again. Is there a way to enforce it even without that restriction I added?

[the remark this answers is not in the document]
Comment (Mark-Simulacrum) [reply src/ci/docker/host-x86_64/x86_64-gnu-llvm-10/Dockerfile:23]: I am saying that we run them by default *if* the needed dependencies are available, but if they are explicitly requested, then they require those dependencies.

[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/ci/docker/host-x86_64/x86_64-gnu-llvm-10/Dockerfile:23]: Makes sense, putting it back then!

**Comment by GuillaumeGomez** [comment]
@Mark-Simulacrum From the last failure, it seems like `x.py test src/test/rustdoc-gui` cannot run because it needs to create files? An idea for this issue maybe?

File: src/ci/docker/host-x86_64/x86_64-gnu-tools/Dockerfile [hunk 13]
```diff
@@ -23,7 +24,17 @@ RUN /scripts/cmake.sh
...
+# to create a new folder. For reference:
+# https://github.com/puppeteer/puppeteer/issues/375
+
 ENV RUST_CONFIGURE_ARGS \
   --build=x86_64-unknown-linux-gnu \
   --save-toolstates=/tmp/toolstate/toolstates.json
-ENV SCRIPT /tmp/checktools.sh ../x.py
+ENV SCRIPT /tmp/checktools.sh ../x.py && \
+  npm install browser-ui-test -g --unsafe-perm=true && \
+  cd .. && python3 ./x.py test src/test/rustdoc-gui --stage 2
```
Comment (Mark-Simulacrum) [inline src/ci/docker/host-x86_64/x86_64-gnu-tools/Dockerfile:40]: This is calling x.py from the current directory, but should be using ../x.py to create the build dir in the right place. We need to support invoking x.py from subdirectories, so paths should always be join'd with builder.config.src or something like that (which points to the root of the rust-lang/rust checkout).

Comment (GuillaumeGomez) [reply src/ci/docker/host-x86_64/x86_64-gnu-tools/Dockerfile:40]: Makes sense. It originally was using `../` (without the the call to `cd`) but it couldn't find `stc/test/rustdoc-gui` so in both cases, it seems like a needed update. I'll send a PR to fix that tomorrow.

**Comment by GuillaumeGomez** [comment]
@Mark-Simulacrum: Review for review! :)

File: src/bootstrap/test.rs [hunk 14]
```diff
@@ -794,7 +807,9 @@ impl Step for RustdocGUI {
     const ONLY_HOSTS: bool = true;
 
     fn should_run(run: ShouldRun<'_>) -> ShouldRun<'_> {
-        run.path("src/test/rustdoc-gui")
+        let builder = run.builder;
+        let run = run.path("src/test/rustdoc-gui");
+        run.default_condition(builder.config.nodejs.is_some())
```
Comment (Mark-Simulacrum) [inline src/bootstrap/test.rs:812]: I believe we had discussed an alternative formulation here:

* use check_if_browser_ui_test_is_installed as the default condition
* if the step is running, panic if anything is not installed (browser-ui-test, node, etc.)

File: src/ci/docker/host-x86_64/x86_64-gnu-tools/Dockerfile [hunk 15]
```diff
@@ -23,7 +63,17 @@ RUN /scripts/cmake.sh
...
   --build=x86_64-unknown-linux-gnu \
   --save-toolstates=/tmp/toolstate/toolstates.json
-ENV SCRIPT /tmp/checktools.sh ../x.py
+
+# Install required dependencies from browser-UI-test framework
+# For now, we need to use `--unsafe-perm=true` to go around an issue when npm tries
+# to create a new folder. For reference:
+# https://github.com/puppeteer/puppeteer/issues/375
+ENV SCRIPT /tmp/checktools.sh ../x.py && \
+  npm install -g browser-ui-test --unsafe-perm=true && \
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/ci/docker/host-x86_64/x86_64-gnu-tools/Dockerfile:78]: Sure!

**Comment by GuillaumeGomez** [comment]
Updated!

**Comment by GuillaumeGomez** [comment]
ping @Mark-Simulacrum

File: src/bootstrap/test.rs [hunk 16]
```diff
@@ -774,6 +774,19 @@ impl Step for RustdocJSNotStd {
     }
 }
 
+fn check_if_browser_ui_test_is_installed(npm: &Path, global: bool) -> bool {
```
Comment (Mark-Simulacrum) [inline src/bootstrap/test.rs:777]: We always call this with global true and false; can you make a wrapper function that calls this twice (with and without global)?

File: src/bootstrap/test.rs [hunk 17]
```diff
@@ -774,6 +774,19 @@ impl Step for RustdocJSNotStd {
...
+    let mut command = Command::new(&npm);
+    command.arg("list").arg("--depth=0");
+    if global {
+        command.arg("--global");
+    }
+    let lines = command
+        .output()
+        .map(|output| String::from_utf8_lossy(&output.stdout).to_string())
```
Comment (Mark-Simulacrum) [inline src/bootstrap/test.rs:785]:
```suggestion
        .map(|output| String::from_utf8_lossy(&output.stdout).into_owned())
```

File: src/bootstrap/test.rs [hunk 18]
```diff
@@ -795,58 +821,59 @@ impl Step for RustdocGUI {
...
+        // We now run GUI tests.
+        let mut command = Command::new(&nodejs);
+        command
+            .arg(builder.build.src.join("src/tools/rustdoc-gui/tester.js"))
+            .arg("--doc-folder")
+            .arg(out_dir)
+            .arg("--tests-folder")
+            .arg(builder.build.src.join("src/test/rustdoc-gui"));
+        if let Ok(node_path) = env::var("NODE_PATH") {
+            command.env("NODE_PATH", node_path);
```
Comment (Mark-Simulacrum) [inline src/bootstrap/test.rs:874]: Why is this necessary? Shouldn't it get inherited by default anyway?

File: src/ci/scripts/should-skip-this.sh [hunk 19]
```diff
@@ -26,6 +26,11 @@ elif ! git diff --quiet "$BASE_COMMIT" -- src/tools/clippy src/tools/rustfmt; th
     # There is not an easy blanket search for subtrees. For now, manually list
     # the subtrees.
     echo "Executing the job since clippy or rustfmt subtree was updated"
+elif (git diff --name-only HEAD^ | grep --quiet src/test/rustdoc-gui) || \
+     (git diff --name-only HEAD^ | grep --quiet src/librustdoc) || \
+     (git diff --name-only HEAD^ | grep --quiet src/tools/rustdoc-gui); then
```
Comment (Mark-Simulacrum) [inline src/ci/scripts/should-skip-this.sh:31]: We've since changed to using the git diff commands with BASE_COMMIT, like above.

File: src/bootstrap/test.rs (hunk 18 above)
Comment (GuillaumeGomez) [reply src/bootstrap/test.rs:874]: No, it needs to be set manually otherwise it's not there (it was why the CI was failing).

Comment (Mark-Simulacrum) [reply src/bootstrap/test.rs:874]: I don't understand -- this is just passing through the environment, it should already be inherited. It presumably needs to be set before calling x.py, but the default is to inherit the environment -- there's no call to env_remove or env_clear here...

File: src/bootstrap/test.rs (hunk 16 above)
Comment (GuillaumeGomez) [reply src/bootstrap/test.rs:777]: Sure!

File: src/bootstrap/test.rs (hunk 18 above)
Comment (GuillaumeGomez) [reply src/bootstrap/test.rs:874]: Hum you're right... I'll give it another try without adding this.

File: src/ci/scripts/should-skip-this.sh (hunk 19 above)
Comment (GuillaumeGomez) [reply src/ci/scripts/should-skip-this.sh]: :+1:

**Comment by GuillaumeGomez** [comment]
@Mark-Simulacrum You were right, it was unneeded to add `NODE_PATH` in the code. Just need to update the tests now then.

**Comment by GuillaumeGomez** [comment]
I'm sending a fix for the last failing test because it is a regression...

**Comment by GuillaumeGomez** [comment]
Test regression fixed in https://github.com/rust-lang/rust/pull/85602. Waiting for it to get merged then...

**Comment by GuillaumeGomez** [comment]
@Mark-Simulacrum Updated!

File: src/ci/scripts/should-skip-this.sh [hunk 20]
```diff
@@ -26,6 +26,11 @@ elif ! git diff --quiet "$BASE_COMMIT" -- src/tools/clippy src/tools/rustfmt; th
     # There is not an easy blanket search for subtrees. For now, manually list
     # the subtrees.
     echo "Executing the job since clippy or rustfmt subtree was updated"
+elif (git diff "$BASE_COMMIT" | grep --quiet src/test/rustdoc-gui) || \
+     (git diff "$BASE_COMMIT" | grep --quiet src/librustdoc) || \
+     (git diff "$BASE_COMMIT" | grep --quiet src/tools/rustdoc-gui); then
```
[the remark this answers is not in the document]
Comment (GuillaumeGomez) [reply src/ci/scripts/should-skip-this.sh]: You're absolutely right!

**Comment by GuillaumeGomez** [comment]
@Mark-Simulacrum Seems like the CI wasn't run on rustdoc-GUI...

**Review by Mark-Simulacrum — APPROVED** [review]
